File: 692.前k个高频单词.py
#
# @lc app=leetcode.cn id=692 lang=python3
#
# [692] 前K个高频单词
#

# @lc code=start
import heapq
class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        dict = {}
        for word in words:
            dict[word] = dict.get(word, 0) + 1
        heap = []
        for key in dict:
            customElement = CustomElement(dict[key], key)
            heapq.heappush(heap, (customElement))
        res = []
        while k > 0:
            element = heapq.heappop(heap)
            res.append(element.val2)
            k -= 1
        return res


class CustomElement:
    def __init__(self, val1, val2) -> None:
        self.val1 = val1
        self.val2 = val2
    
    def __lt__(self, other):
        if self.val1 == other.val1:
            return self.val2 < other.val2
        return self.val1 > other.val1
# @lc code=end

File: 692.前k个高频单词.java
/*
 * @lc app=leetcode.cn id=692 lang=java
 *
 * [692] 前K个高频单词
 */

// @lc code=start
import java.util.ArrayList;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.List;
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        HashMap<String, Integer> map = new HashMap<>();
        for(String word: words){
            map.put(word, map.getOrDefault(word, 0)+1);
        }
        PriorityQueue<WordFreq> queue = new PriorityQueue<>();
        for(String key:map.keySet()){
            WordFreq wordFreq = new WordFreq(map.get(key), key);
            queue.offer(wordFreq);
        }
        List<String> list = new ArrayList<>();
        while(k>0){
            list.add(queue.poll().word);
            k--;
        }
        return list;

    }

    public static class WordFreq implements Comparable<WordFreq>{
        int freq;
        String word;
        
        WordFreq(int freq, String word){
            this.freq = freq;
            this.word = word;
        }

        public int compareTo(WordFreq other){
            if(this.freq == other.freq){
                return this.word.compareTo(other.word);
            }
            return other.freq - this.freq;
        }
    }
}
// @lc code=end

